torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -137,74 +137,75 @@ std::unique_ptr<RpcCommandBase> RequestCallbackImpl::
   return pythonRpc ? std::move(pythonRpc) : std::move(rpc);
 }
 
-void RequestCallbackImpl::processScriptCall(
-    RpcCommandBase& rpc,
-    const std::function<void(Message)>& markComplete,
-    const c10::intrusive_ptr<JitFuture>& responseFuture) const {
+c10::intrusive_ptr<JitFuture> RequestCallbackImpl::processScriptCall(
+    RpcCommandBase& rpc) const {
   auto& scriptCall = static_cast<ScriptCall&>(rpc);
-  auto& stack = scriptCall.stackRef();
+
+  c10::intrusive_ptr<JitFuture> future;
   if (scriptCall.hasOp()) {
-    processScriptCallOp(scriptCall, markComplete, stack);
-    return;
+    future = runJitOperator(*scriptCall.op(), scriptCall.stackRef());
+  } else {
+    future = runJitFunction(
+        scriptCall.qualifiedName(),
+        scriptCall.stackRef(),
+        scriptCall.isAsyncExecution());
   }
 
-  auto jitFuture = runJitFunction(
-      scriptCall.qualifiedName(), stack, scriptCall.isAsyncExecution());
-
-  jitFuture->addCallback([responseFuture, markComplete](JitFuture& jitFuture) {
-    if (jitFuture.hasError()) {
-      responseFuture->setError(jitFuture.exception_ptr());
-    } else {
-      responseFuture->markCompleted(c10::make_intrusive<Message>(
-          ScriptResp(jitFuture.value()).toMessage()));
-    }
-  });
+  return future->then(
+      [](JitFuture& jitFuture) {
+        return c10::make_intrusive<Message>(
+            ScriptResp(jitFuture.value()).toMessage());
+      },
+      c10::getCustomClassType<c10::intrusive_ptr<Message>>());
 }
 
-void RequestCallbackImpl::processPythonCall(
-    RpcCommandBase& rpc,
-    const std::function<void(Message)>& markComplete,
-    const c10::intrusive_ptr<JitFuture>& responseFuture) const {
+c10::intrusive_ptr<JitFuture> RequestCallbackImpl::processPythonCall(
+    RpcCommandBase& rpc) const {
   auto& upc = static_cast<UnpickledPythonCall&>(rpc);
   auto future = runPythonFunction(upc.pythonUdf(), upc.isAsyncExecution());
 
-  future->addCallback([responseFuture](JitFuture& future) {
-    auto& pythonRpcHandler = PythonRpcHandler::getInstance();
-    py::gil_scoped_acquire acquire;
-    auto serializedPyObj =
-        pythonRpcHandler.serialize(jit::toPyObject(future.value()));
-    responseFuture->markCompleted(c10::make_intrusive<Message>(
-        PythonResp(std::move(serializedPyObj)).toMessage()));
-  });
+  return future->then(
```

Comment:
Each of the futures in the chain will introduce another data ptrs extraction for CUDA (when CUDA is enabled here). Do we need to worry about the perf issue here?

Reply:
I have been asking myself the same thing. I'm somewhat worried about DataPtr extraction (although not too much, because for the most part here we're inspecting JIT IValues, which should be much cheaper than inspecting Python objects) but I'm also a bit worried about creating/recording/waiting on events each time, getting new streams, ...

I have some ideas on how we can mitigate some of these issues. For example in #58424 I'll allow callbacks to provide pre-extracted DataPtrs, which should allow us to skip extraction in some cases. I also believe we can add some fastpaths in Future so that for example we reuse the same streams from the "parent" when running callbacks.

Honestly, though, I'm also fine for now to take a perf hit if necessary, if that means we can get CUDA support to work correctly and reliably, and then look later into optimizing it.